sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -169,16 +179,22 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
             }
           }
 
+          val (dataSchema, updatedTable) = inferIfNeeded(
+            relation,
+            options,
+            filteredMetastoreSchema,
+            fileFormat,
+            Option(fileIndex))
+
           val fsRelation = HadoopFsRelation(
             location = fileIndex,
             partitionSchema = partitionSchema,
-            dataSchema = relation.tableMeta.dataSchema,
+            dataSchema = dataSchema,
             // We don't support hive bucketed tables, only ones we write out.
             bucketSpec = None,
-            fileFormat = fileFormatClass.newInstance(),
+            fileFormat = fileFormat,
             options = options)(sparkSession = sparkSession)
-
-          val created = LogicalRelation(fsRelation, catalogTable = Some(relation.tableMeta))
+          val created = LogicalRelation(fsRelation, catalogTable = Option(updatedTable))
```

Comment:
I don't think I was very clear. Hypothetically, if ```a``` were null when it shouldn't be:

1. ```Some(a)``` will result in ```Some(null)```
2. ```Option(a)``` will result in ```None```

Based on your response, (1) is the preferred approach, meaning a runtime error will occur when attempting to use the (expected non-null) value rather than just thinking an optional value wasn't provided.